### PR TITLE
Added UT for reconcilePowerVSResources function

### LINKED ISSUE #1818: EPIC: Add UT accross various files in repository
/kind feature
/area provider/ibmcloud

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]

Recently we added lots of functionality but haven't added any corresponding UT's to it. Its time to add unit test cases and overall goal is to reach 80% of test coverage.

As a non-goal during the time of adding UT if we find any redundant or unwanted or code duplication, a separate PR needs to be submitted to optimize the code.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

- [ ] cloud/scope/powervs_cluster.go 
- [ ] cloud/scope/powervs_machine.go 

Instructions for running individual tests on VS Code

1. Get kubebuilder asset path
```
$ cd cluster-api-provider-ibmcloud
$ make setup-envtest
kube-builder assets set using darwin OS at location /Users/karthikkn/Library/Application Support/io.kubebuilder.envtest/k8s/1.29.3-darwin-amd64
```

2. Create settings.json file with kubebuilder asset path
```
$ cd cluster-api-provider-ibmcloud/.vscode
$ cat settings.json
cat settings.json
{
    "go.testEnvVars": {
        "KUBEBUILDER_ASSETS": "/Users/karthikkn/Library/Application Support/io.kubebuilder.envtest/k8s/1.29.3-darwin-amd64",
    },
}

```
3. Run individual test by clicking on run test above test function name in VS Code editor

Instructions for checking code coverage of file 

1. Generate the test coverage html page
```
$ cd cluster-api-provider-ibmcloud/.vscode
$ make test-cover
```
A file named cover.html will be created under cluster-api-provider-ibmcloud directory

example path: /Users/karthikkn/k8-workspace/cluster-api-provider-ibmcloud/cover.html

2. Copy the path of the file and paste in browser to view the html page and select the file to view the lines covered in UT.

image

```[tasklist]
### Helper PRs
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1913
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1899
```

```[tasklist]
### UT PRs
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1912
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1927
- [ ] https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1928
```

### PR DESCRIPTION
**Pull request by Shilpa-Gokul**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially Fixes #1818 

**Special notes for your reviewer**:

/area provider/ibmcloud

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

### CODE REVIEW

File: controllers/ibmpowervscluster_controller_test.go [hunk 1]
```diff
@@ -456,3 +465,174 @@ func cleanupCluster(g *WithT, powervsCluster *infrav1beta2.IBMPowerVSCluster, na
...
+					Status:             "False",
+					Severity:           capiv1beta1.ConditionSeverityError,
+					LastTransitionTime: metav1.Time{},
+					Reason:             infrav1beta2.ServiceInstanceReconciliationFailedReason,
+					Message:            "error getting resource instance",
+				},
+			},
+		},
+		{
+			name: "when Reconciling PowerVS service instance returns requeue",
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:509]:
```suggestion
			name: "when Reconciling PowerVS service instance returns requeue as  true",
```
Please change wherever applicable

File: controllers/ibmpowervscluster_controller_test.go [hunk 2]
```diff
@@ -456,3 +465,174 @@ func cleanupCluster(g *WithT, powervsCluster *infrav1beta2.IBMPowerVSCluster, na
...
+				return clusterScope
+			},
+			reconcileResult: reconcileResult{
+				Result: reconcile.Result{
+					Requeue: true,
+				},
+			},
+		},
+		{
+			name: "when Reconciling network returns requeue",
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:532]:
```suggestion
			name: "when Reconciling network returns error",
```

File: controllers/ibmpowervscluster_controller_test.go [hunk 3]
```diff
@@ -456,3 +465,174 @@ func cleanupCluster(g *WithT, powervsCluster *infrav1beta2.IBMPowerVSCluster, na
...
+					Severity:           capiv1beta1.ConditionSeverityError,
+					LastTransitionTime: metav1.Time{},
+					Reason:             infrav1beta2.NetworkReconciliationFailedReason,
+					Message:            "DHCP server creation failed and is in error state",
+				},
+				getServiceInstanceReadyCondition(),
+			},
+		},
+		{
+			name: "when dhcpServer is active",
```
Comment (Amulyam24) [inline controllers/ibmpowervscluster_controller_test.go:571]:
```suggestion
			name: "when reconcile network returns with DHCP server in active state",
```

**Review by Amulyam24 — COMMENTED** [review]
Minor suggestions, overall LGTM!

**Review by Amulyam24 — APPROVED** [review]
/lgtm

**Comment by Karthik-K-N** [comment]
/hold
#2032 should merge first as its huge, so This PR needs a rebase after that.

**Comment by Shilpa-Gokul** [comment]
@Karthik-K-N @Amulyam24 Have rebased the changes, PTAL

File: controllers/ibmpowervscluster_controller_test.go [hunk 4]
```diff
@@ -1146,19 +1146,6 @@ func TestReconcileVPCResources(t *testing.T) {
...
-	}
-}
-func getNetworkReadyCondition() capiv1beta1.Condition {
-	return capiv1beta1.Condition{
-		Type:   infrav1beta2.NetworkReadyCondition,
-		Status: "True",
-	}
-}
-
 func getVPCReadyCondition() capiv1beta1.Condition {
```
Comment (Karthik-K-N) [inline controllers/ibmpowervscluster_controller_test.go:1149]: Lets move all test helper functions to last(after all the test functions)

File: controllers/ibmpowervscluster_controller_test.go [hunk 5]
```diff
@@ -1328,3 +1315,174 @@ func cleanupCluster(g *WithT, powervsCluster *infrav1beta2.IBMPowerVSCluster, na
...
+					IBMPowerVSCluster: &infrav1beta2.IBMPowerVSCluster{
+						Status: infrav1beta2.IBMPowerVSClusterStatus{
+							ServiceInstance: &infrav1beta2.ResourceReference{
+								ID: ptr.To("serviceInstanceID"),
+							},
+						},
+					},
+				}
+				mockResourceController := resourceclientmock.NewMockResourceController(gomock.NewController(t))
+				mockResourceController.EXPECT().GetResourceInstance(gomock.Any()).Return(&resourcecontrollerv2.ResourceInstance{State: (*string)(&infrav1beta2.ServiceInstanceStateProvisioning), Name: ptr.To("serviceInstanceName")}, nil, nil)
```
Comment (Karthik-K-N) [inline controllers/ibmpowervscluster_controller_test.go:1371]: Can we get State like this ```ptr.To(string(infrav1beta2.ServiceInstanceStateProvisioning))```
(here and everywhere its applicable)

File: controllers/ibmpowervscluster_controller_test.go [hunk 6]
```diff
@@ -1146,16 +1146,161 @@ func TestReconcileVPCResources(t *testing.T) {
...
+			pvsCluster := &powerVSCluster{
+				cluster: clusterScope.IBMPowerVSCluster,
+			}
+			wg := &sync.WaitGroup{}
+			wg.Add(1)
+			reconciler.reconcilePowerVSResources(clusterScope, pvsCluster, ch, wg)
+			wg.Wait()
+			close(ch)
+			g.Expect(<-ch).To(Equal(tc.reconcileResult))
+			if len(tc.conditions) > 0 {
```
Comment (Karthik-K-N) [inline controllers/ibmpowervscluster_controller_test.go:1297]: See even this condition might not required, Forgot to mention previosly

File: controllers/ibmpowervscluster_controller_test.go [hunk 7]
```diff
@@ -1146,16 +1146,161 @@ func TestReconcileVPCResources(t *testing.T) {
...
-		Status: "True",
+func TestReconcilePowerVSResources(t *testing.T) {
+	testCases := []struct {
+		name                    string
+		powerVSClusterScopeFunc func() *scope.PowerVSClusterScope
+		reconcileResult         reconcileResult
+		conditions              capiv1beta1.Conditions
+	}{
+		{
+			name: "when Reconciling PowerVS service instance returns error",
```
Comment (Karthik-K-N) [inline controllers/ibmpowervscluster_controller_test.go:1157]: May be for being consistent with other files, lets start with Capital letters for Name!

**Review by Prajyot-Parab — APPROVED** [review]
/lgtm

**Pull request merged**